### ISSUE TITLE
Make reduceRight lazy

### ISSUE THREAD

**Issue by wshager** [facebook/immutable-js#981]
It has become clear to me that `Seq.reduce` is not lazy. Currently reduceRight is simply the reverse of reduce. However, I read that in reduceRight _can_ be lazy: https://en.wikipedia.org/wiki/Fold_(higher-order_function)#Evaluation_order_considerations. Perhaps it is worth while to implement this.

**Comment by i-am-tom** [comment]
This concept works really well for languages like Haskell where recursion and lazy evaluation are built-in principles, but isn't quite so effective for languages without such optimisation. To write a (crude) right fold in JS:

```
reduceRight = (f, acc, [x, ... xs]) => x == undefined ? f(x, reduceRight(f, acc, xs)) : acc
```

In the `f(x, reduceRight(f, acc, xs))` call, the entire remainder of the reduction must be calculated to do this step because parameters are always evaluated before calling. You could get round this with some generator trickery, *but* it isn't guaranteed that your folding function lends itself to laziness, so this isn't going to work in all cases and certainly won't look much like a right fold.

tl;dr - right reduction is _super_ cool in lazy languages, and not so cool in eager languages.

**Comment by wshager** [comment]
So... why is it there?

I saw in Clojure they introduced 'reduced'. I guess you could add the same trick in JS, by evaluating an extra check function and exiting when it returns true. I don't see how a generator function could help, since that's already your input, right?

**Comment by leebyron** [comment]
I think it's more important for `reduce` and `reduceRight` to mirror the behavior of the same named functions on `Array` for predictability.

If anyone wanted to submit PRs to add laziness to reduce/reduceRight, that would certainly be interesting but it would either need to be a different name (fold?) or not upset existing predictable behavior.

**Comment by wshager** [comment]
@leebyron fair enough. The proposed kind of lazyness is outside the JS paradigm in any case, as it would yield values that need to be processed in a specific way.
